### ISSUE TITLE
[REVIEW] Compile documentation with Sphinx

### ISSUE THREAD

**Issue by santisoler** [underworldcode/stripy#28]
Part of https://github.com/openjournals/joss-reviews/issues/1410

I've found that the `stripy` public functions and classes are very well documented, and together with the Notebooks they give a nice starting point for newcomers.
So I consider the **Example usage** item on the review checklist is already satisfied with the notebooks, while the **Functionality documentation** is covered with the functions and classes docstrings.
Nevertheless, reading documentation from source code files or asking a specific docstring from a Jupyter Notebook can be uncomfortable on the long term.

In sumary, I think both items are **already satisfied**, but they could be improved with a compiled documentation.
Building and uploading the package documentation to a website is a best practice for software development, making a fast a very helpful way to learn how to use its functions and classes.

One of the most used tools to do so is [Sphinx](https://www.sphinx-doc.org/en/stable/).
It can automatically generate an HTML version of your package documentation by automatically compiling your Python functions' docstrings.
So, you just need to configure Sphinx and not worry about copying and pasting documentation.
Besides, an elegant HTML documentation is a very nice start point for newcomers.

A lot of packages use Sphinx for building their documentation. Just for citing some of them, see [verde](https://github.com/fatiando/verde) or [yellowbrick](https://github.com/districtdatalabs/yellowbrick).

The files for building the documentation are usually under the `docs` directory.
After Sphinx is run, it generates the website as HTML files, which could be uploaded to be available online.
This task can be automated by using [Travis CI](https://travis-ci.org/) or any other continuous integration service (Travis CI is completely free of charge for GitHub public repos).
So for example, you could have a `docs` directory on your repo (under the `master` branch), configure Travis CI to build a new version of the website after a new commit on `master` is pushed and then make Travis to deploy the HTML files to a `gh_pages` branch inside the same repo.
Github offers you the option to serve a static website on any specific branch (also for free), so in this case you should choose `gh_pages`.
In conclusion, your documentation will be automatically generated and uploaded to https://underworldcode.github.io/stripy making it publicly available.

If you consider that generating a documentation would be helpful and want to implement it, let me know.
I'm willing to help you to tackle this.

**Comment by lmoresi** [comment]
This is probably a good idea, but I wonder if it may be better to do this after the formal review and ask for your help if you are still able to find the time. I would prefer that we give you credit for that level of assistance that is beyond simply a review.

**Comment by santisoler** [comment]
Sounds good to me, we could leave the documentation compilation for the future and focus on more *urgent* issues.

Please, leave this issue open to remind us that we want to tackle this on the future.

PS: Thanks for your consideration about my assistance. I think the open review opens the possibility to create these kind of contributions between people around the world. I'm glad I can help and experience this very nice way to perform peer review.

**Comment by brmather** [comment]
While this is independent of the formal review, I was interested to check out how this compiled documentation works. The workflow @santisoler suggests looks sound, but I encountered some issues with `sphinx`:

- `sphinx` pollutes the `docs` folder with makefiles, configuration files, etc. which I find messy.
- Markdown is not really supported and it is a pain to format everything to restructured text since everything is in markdown these days. The `recommonmark` package tries to fix this but does not *quite* cut it in my view.
- No matter how much I beat the docstrings to conform to numpy or google formatting, there were many instances where I simply could not get it to display correctly.
- Generally, I find the learning curve is quite high to make decent looking docs that render correctly.

Alternatives like `mkdocs` don't automatically compile docstrings, so that's out, but I stumbled across [`pdoc3`](https://pdoc3.github.io/pdoc/) which was a refreshingly pleasant experience. I only had to make very minimal changes to the docstrings to get it to format correctly and it requires no configuration files.

Just run:

```sh
pdoc3 --html -o docs/ stripy
```

and it compiles all the docstrings and shoves the HTML in your `docs` directory. I've been working off the [compiled-documentation](https://github.com/underworldcode/stripy/tree/compiled-documentation) branch if anyone want to take `pdoc3` for a spin. Remember to compile and install `stripy` from that branch first before attempting this as `pydoc3` looks up the installed version in the system/user directory to compile docstrings.

**Comment by lmoresi** [comment]
The benefit seems to be that, like pytest, it is very lightweight and adds little overhead. Their motto is also one of the better ones out there.

L

Prof Louis Moresi

louis.moresi@unimelb.edu.au<mailto:louis.moresi@unimelb.edu.au>

(w) +61 3 8344 1217

(m) +61 4 0333 1413

(us) +1 505 349 4425

www.moresi.info<http://www.moresi.info/>

www.facebook.com/underworldcode<http://www.facebook.com/underworldcode>

@LouisMoresi<https://twitter.com/LouisMoresi>


On 3 Jun 2019, 6:43 PM +1000, Ben Mather <notifications@github.com>, wrote:

While this is independent of the formal review, I was interested to check out how this compiled documentation works. The workflow @santisoler<https://github.com/santisoler> suggests looks sound, but I encountered some issues with sphinx:

  *   sphinx pollutes the docs folder with makefiles, configuration files, etc. which I find messy.
  *   Markdown is not really supported and it is a pain to format everything to restructured text since everything is in markdown these days. The recommonmark package tries to fix this but does not quite cut it in my view.
  *   No matter how much I beat the docstrings to conform to numpy or google formatting, there were many instances where I simply could not get it to display correctly.
  *   Generally, I find the learning curve is quite high to make decent looking docs that render correctly.

Alternatives like mkdocs don't automatically compile docstrings, so that's out, but I stumbled across pdoc3<https://pdoc3.github.io/pdoc/> which was a refreshingly pleasant experience. I only had to make very minimal changes to the docstrings to get it to format correctly and it requires no configuration files.

Just run:

pdoc3 --html -o docs/ stripy

and it compiles all the docstrings and shoves the HTML in your docs directory. I've been working off the compiled-documentation<https://github.com/underworldcode/stripy/tree/compiled-documentation> branch if anyone want to take pdoc3 for a spin. Remember to compile and install stripy from that branch first before attempting this as pydoc3 looks up the installed version in the system/user directory to compile docstrings.

**Comment by brmather** [comment]
> Their motto is also one of the better ones out there.

:laughing: 

@lmoresi can you set up the GitHub pages environment? I would do it but I don't have access to the settings panel for the repo...

**Comment by santisoler** [comment]
> While this is independent of the formal review, I was interested to check out how this compiled documentation works.

Nice initiative! Documentation is a very important part of releasing software. Helping users to figure out how it works always makes things easy for them.

> The workflow @santisoler suggests looks sound, but I encountered some issues with `sphinx`:
> 
>     * `sphinx` pollutes the `docs` folder with makefiles, configuration files, etc. which I find messy.

Your right, Sphinx builds the docs upon configuration files and a `Makefile`. Could be a little bit overwhelming at first, I agree.
Although, I think the `Makefile` is great when you need to tell a CI (continuous integration) to build the documentation: with a simple `make` command you ensure that the CI will perform the same task you are doing when building the documentation locally. Moreover, if some argument for the building has to be changed you just do it on the `Makefile`, leaving the CI configuration file unmodified.

>     * Markdown is not really supported and it is a pain to format everything to restructured text since everything is in markdown these days. The `recommonmark` package tries to fix this but does not _quite_ cut it in my view.

You're also right about this. It's a little bit annoying to use RST, Markdown is way simpler.
Nevertheless, you need to edit these files only once, or maybe when you make a new release, include a new contributor, etc. But I get it could be annoying.

>     * No matter how much I beat the docstrings to conform to numpy or google formatting, there were many instances where I simply could not get it to display correctly.

I always follow the [numpydoc docstring guide](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) when I struggle with these issues.
Nevertheless, the `stripy` docstrings are very well written and after a quick view I think they are in great agreement with the numpydoc docstring guide.
Maybe what's wrong is the single space before each parameter or returns. For example, I'll copy the `sTriangulation` docstring. I replaced these spaces with an underscore:
```
    Parameters
    ----------
    _lons    : 1D array of longitudinal coordinates in radians
    _lats    : 1D array of latitudinal coordinates in radians
    _refinement_levels (int)
    _        : refine the number of points in the triangulation
    _          (see uniformly_refine_triangulation)
    _permute : (bool) randomises the order of lons and lats to improve
    _          triangulation efficiency and eliminate colinearity
    _          issues (see notes)
    _tree    : (bool) construct a cKDtree for efficient nearest-
               neighbour lookup
```

I'm curious if that's the case, but don't worry if you don't want to test it. I can live with this doubt haha.

>     * Generally, I find the learning curve is quite high to make decent looking docs that render correctly.

Maybe you're right about this. That's why I haven't asked you to tackle down this issue in order to get the publication accepted.
From my experience, having an automated documentation that looks very nice afterwards worth the effort, but I agree that Sphinx could be a little bit difficult. 
I'm glad it's not the only way to solve this.

> Alternatives like `mkdocs` don't automatically compile docstrings, so that's out, but I stumbled across [`pdoc3`](https://pdoc3.github.io/pdoc/) which was a refreshingly pleasant experience. I only had to make very minimal changes to the docstrings to get it to format correctly and it requires no configuration files.
> 
> Just run:
> 
> ```shell
> pdoc3 --html -o docs/ stripy
> ```
> 
> and it compiles all the docstrings and shoves the HTML in your `docs` directory. I've been working off the [compiled-documentation](https://github.com/underworldcode/stripy/tree/compiled-documentation) branch if anyone want to take `pdoc3` for a spin. Remember to compile and install `stripy` from that branch first before attempting this as `pydoc3` looks up the installed version in the system/user directory to compile docstrings.

That's great! I have never used `pdoc` but seems a very nice and easy to use tool.
It's very well suited for `stripy` and the `pdoc` documentation (built by itself) looks very nice!
I can't wait to see how `stripy` documentation looks with `pdoc`.

Great job @brmather!

**Comment by lmoresi** [comment]
The docs are live ... www.underworldcode.org/stripy (or underworldcode.github.io/stripy. 

The one thing that would be good is a solution to the problem of embedding equations in the docstrings. It is possible with sphinx and it is somewhat possible with pdoc3 but it complains about the latex backslash escape characters which it is trying to interpret. 

However, take a look. It seems to be at the right level for the project and, I must say, it does also make it helpful in keeping an eye on whether we are correctly contructing and populating the docstrings.

**Comment by brmather** [comment]
Ah, be sure to compile off the "compiled-documentation" branch. I made a few minor adjustment to the docstrings to get them to render correctly (see pull request #37).

**Comment by brmather** [comment]
OK the [docs](http://www.underworldcode.org/stripy/) are looking *very, very* good!

> The one thing that would be good is a solution to the problem of embedding equations in the docstrings. It is possible with sphinx and it is somewhat possible with pdoc3 but it complains about the latex backslash escape characters which it is trying to interpret.

I'm not sure how to embed maths expressions, however, the backslash shouldn't be a problem if you put *r* in front of the docstring, right? If not, it might be worth raising an [issue ](https://github.com/pdoc3/pdoc/issues) on their GitHub since `pdoc` is under active development.

**Comment by lmoresi** [comment]
https://github.com/lmoresi/pdoc - I made a quick fork and added some mathjax support and python markdown support. The issue of escaping still arises which can be avoided by escaping the escapes or by marking the docstring as raw. There are some deprecated features in the markdown math extension but ... details. 

It's not obvious which is the right choice but I am not that keen on the raw marker as it looks horrible. I am also not sure if it would have side effects to coerce all docstrings to be interpreted as raw when pdoc launches. 

Have a look though, because there are quite a few places where it would be useful to introduce mathematical notation (e.g. in the derivatives). We can ask the pdoccers about a PR when we have a working example posted ... see below for a first example

http://www.underworldcode.org/stripy/cartesian.html#stripy.cartesian.Triangulation.gradient

**Comment by brmather** [comment]
This seems pretty good. It supports fractions and symbols e.g. [`interpolate_cubic()`](http://www.underworldcode.org/stripy/cartesian.html#stripy.cartesian.Triangulation.interpolate_cubic). The MathJax(ed) docs look decent for `cartesian.py` and it's only a matter of time before `spherical.py` looks as good!

**Comment by lmoresi** [comment]
Nice … I wonder if \partial F / \partial y looks better than the \frac equivalent for inline stuff.

Prof Louis Moresi

louis.moresi@unimelb.edu.au<mailto:louis.moresi@unimelb.edu.au>

(w) +61 3 8344 1217

(m) +61 4 0333 1413

(us) +1 505 349 4425

www.moresi.info<http://www.moresi.info/>

www.facebook.com/underworldcode<http://www.facebook.com/underworldcode>

@LouisMoresi<https://twitter.com/LouisMoresi>


On 4 Jun 2019, 4:49 PM +1000, Ben Mather <notifications@github.com>, wrote:

This seems pretty good. It supports fractions and symbols e.g. interpolate_cubic()<http://www.underworldcode.org/stripy/cartesian.html#stripy.cartesian.Triangulation.interpolate_cubic>. The MathJax(ed) docs look decent for cartesian.py and it's only a matter of time before spherical.py looks as good!

**Comment by santisoler** [comment]
:tada: Nice job! :tada:

The documentation looks really good!
I liked how pdoc created a very beautiful documentation with a very low learning curve.
Congratulations! :muscle: 

One minor thought.
Have you compiled the documentation locally and then pushed it to the repo?
This approach is very quick but might be annoying in the future: each time you make some change on the code, you need to remember to update the documentation.
Another approach would be to configure a CI (Travis CI is free for public Github repositories) to compile the documentation and push it to a dedicated branch, e.g. `gh-pages`. That branch will be only handled by the CI, which will update it when a new commit is pushed to `master`.
From my experience, working with CIs makes maintenance a lot easier and allow us to focus more on the most important part of the repository (design, code, how docstrings are written, etc).

The [Travis CI documentation](https://docs.travis-ci.com/) is very easy to understand. To configure it you'll just need to write a `.travis.yml` file, and on this specific case you may need some scripts for deploying the documentation into the `gh-pages` branch.
@leouieda wrote some continuous integration scripts on [fatiando/continuous-integration](https://github.com/fatiando/continuous-integration). Some of them are [specifically for Travis](https://github.com/fatiando/continuous-integration/tree/master/travis).
I always refer to these scripts when I need to configure Travis and/or deploy from it.

Just want to make your life easier with this advice.
I think I'll spend a few more minutes appreciating the new Stripy documentation! :relaxed:

**Comment by santisoler** [comment]
I've just seen you started configuring Travis a few hours ago! That's great!
Hope my previous message helps.

**Comment by brmather** [comment]
Yes @santisoler I had the same thought! Over time it would get very tedious to manually compile the documentation. @lmoresi I don't think the repo is configured to let me manage `stripy` on Travis ([travis-ci.org](https://travis-ci.org) or [travis-ci.com](https://travis-ci.com/)). In any case, I tried it out my [`pycurious`](https://github.com/brmather/pycurious) package and it works really well. I've copied the `.travis.yml` file to `stripy` so setting it up should be relatively straightforward. Here are the steps I took:

1. Setup `stripy` in Travis CI with a [personal access token](https://github.com/settings/tokens)
2. Create a *gh-pages* branch for Travis to push the documentation. You can also configure it to go in the `docs` folder, but I find this messy because Travis ends up committing to the master branch, and the HTML pages inside `docs` don't render in GitHub anyway - so I'd prefer if that folder was deleted.
3. Configure the `.travis.yml` setup (skip this because I've already done it)
4. Grab a gin & tonic because that's it! :cocktail:

I haven't configured Travis to run pytest on all the test scripts because some of those optional packages that are imported in `test_0_imports.py` would be tricky to setup inside the environment (I'm looking at you `lavavu`, `gdal` and `cartopy`! :eyes: ) So we should make a decision whether those tests ought to stay...

**Comment by brmather** [comment]
@santisoler I just looked those Travis CI scripts on fatiando - the one on deploying to PyPI is super useful! Thanks :+1:

**Comment by santisoler** [comment]
Glad to be helpful!

> In any case, I tried it out my pycurious package and it works really well. I've copied the .travis.yml file to stripy so setting it up should be relatively straightforward.

Nice package you have! I'll leave a comment about it on its repo.

> 1. Setup `stripy` in Travis CI with a [personal access token](https://github.com/settings/tokens)
> 
> 1. Create a _gh-pages_ branch for Travis to push the documentation. You can also configure it to go in the `docs` folder, but I find this messy because Travis ends up committing to the master branch, and the HTML pages inside `docs` don't render in GitHub anyway - so I'd prefer if that folder was deleted.
> 
> 1. Configure the `.travis.yml` setup (skip this because I've already done it)
> 
> 1. Grab a gin & tonic because that's it! :cocktail:

That's great! Specially the last step haha!
Be careful with the personal access token! You **must encrypt** it and prevent Travis from dumping it into the Job Log (see [this line](https://github.com/fatiando/continuous-integration/blob/master/travis/deploy-gh-pages.sh#L32)).
If you have the slightly doubt that the access token has been compromised (unencrypted published or dumped by Travis on the logs), you should delete it and create a new one.

> I haven't configured Travis to run pytest on all the test scripts because some of those optional packages that are imported in `test_0_imports.py` would be tricky to setup inside the environment (I'm looking at you `lavavu`, `gdal` and `cartopy`! :eyes: ) So we should make a decision whether those tests ought to stay...

I normally use `conda` for managing python packages and dependencies.
I had no problem installing `gdal` or `cartopy` from the `conda-forge`, so that may help.
An easy way to get `conda` installed in Travis is through the [`setup-miniconda.sh`](https://github.com/fatiando/continuous-integration/blob/master/travis/setup-miniconda.sh) script.
You could create an `environment.yml` detailing the dependencies for Stripy and telling `conda` to create an environment with those dependencies (`conda` will create the environment, download and install all the dependencies).
If no package is available on `conda-forge` you could alternatively install it through `pip`.
Buuuut, I haven't been able to install `lavavu` from `pip`.

So, I think is not to hard to make Travis run pytest, but you will encounter some issues with `lavavu`.

I'm really happy with the shape Stripy is taking. Sorry to be repetitive, but I'm glad with all the effort you are putting on this. You earned that gin & tonic! Cheers!

**Comment by lmoresi** [comment]
I’d suggest we consider docker here. Should be reasonable to have a stable base image and build against that. This would be advantageous given our reliance on docker for nearly every other workflow - and because of the complexity of our dependencies.

This would be a good thing for other packages and deployments (e.g. of courses) that contain underworld dependencies where the setup is fragile and murderously difficult.

Conda is not a friend to me. I have all sorts of scars from trying to make it do what I want in the broader environment.


Prof Louis Moresi

louis.moresi@unimelb.edu.au<mailto:louis.moresi@unimelb.edu.au>

(w) +61 3 8344 1217

(m) +61 4 0333 1413

(us) +1 505 349 4425

www.moresi.info<http://www.moresi.info/>

www.facebook.com/underworldcode<http://www.facebook.com/underworldcode>

@LouisMoresi<https://twitter.com/LouisMoresi>


On 4 Jun 2019, 11:45 PM +1000, Santiago Soler <notifications@github.com>, wrote:

Glad to be helpful!

In any case, I tried it out my pycurious package and it works really well. I've copied the .travis.yml file to stripy so setting it up should be relatively straightforward.

Nice package you have! I'll leave a comment about it on its repo.

  1.  Setup stripy in Travis CI with a personal access token<https://github.com/settings/tokens>

  2.  Create a gh-pages branch for Travis to push the documentation. You can also configure it to go in the docs folder, but I find this messy because Travis ends up committing to the master branch, and the HTML pages inside docs don't render in GitHub anyway - so I'd prefer if that folder was deleted.

  3.  Configure the .travis.yml setup (skip this because I've already done it)

  4.  Grab a gin & tonic because that's it! 🍸

That's great! Specially the last step haha!
Be careful with the personal access token! You must encrypt it and prevent Travis from dumping it into the Job Log (see this line<https://github.com/fatiando/continuous-integration/blob/master/travis/deploy-gh-pages.sh#L32>).
If you have the slightly doubt that the access token has been compromised (unencrypted published or dumped by Travis on the logs), you should delete it and create a new one.

I haven't configured Travis to run pytest on all the test scripts because some of those optional packages that are imported in test_0_imports.py would be tricky to setup inside the environment (I'm looking at you lavavu, gdal and cartopy! 👀 ) So we should make a decision whether those tests ought to stay...

I normally use conda for managing python packages and dependencies.
I had no problem installing gdal or cartopy from the conda-forge, so that may help.
An easy way to get conda installed in Travis is through the setup-miniconda.sh<https://github.com/fatiando/continuous-integration/blob/master/travis/setup-miniconda.sh> script.
You could create an environment.yml detailing the dependencies for Stripy and telling conda to create an environment with those dependencies (conda will create the environment, download and install all the dependencies).
If no package is available on conda-forge you could alternatively install it through pip.
Buuuut, I haven't been able to install lavavu from pip.

So, I think is not to hard to make Travis run pytest, but you will encounter some issues with lavavu.

I'm really happy with the shape Stripy is taking. Sorry to be repetitive, but I'm glad with all the effort you are putting on this. You earned that gin & tonic! Cheers!

**Comment by brmather** [comment]
Yes fair enough, we already run the tests in the Docker build using `pytest` so no need to double up with Travis CI.

**Comment by lmoresi** [comment]
I meant that travis has a whole workflow that includes docker and we could make use of that. The docker build / test does not give great warnings.

Prof Louis Moresi

louis.moresi@unimelb.edu.au<mailto:louis.moresi@unimelb.edu.au>

(w) +61 3 8344 1217

(m) +61 4 0333 1413

(us) +1 505 349 4425

www.moresi.info<http://www.moresi.info/>

www.facebook.com/underworldcode<http://www.facebook.com/underworldcode>

@LouisMoresi<https://twitter.com/LouisMoresi>


On 5 Jun 2019, 9:40 AM +1000, Ben Mather <notifications@github.com>, wrote:

Yes fair enough, we already run the tests in the Docker build using pytest so no need to double up with Travis CI.

**Comment by santisoler** [comment]
Configuring Travis to run the tests from within your Docker image would be nice, specially if that's your best recommended way to run Stripy.

My conda proposal was just to run the tests for the non-Docker installation. You could either use the standard Python distribution included on Travis CI VMs, but @brmather exposed some issues regarding a few dependencies (`gdal`, `cartopy`, `lavavu`). That's why I offered the conda alternative.

In summary, I agree with @lmoresi: running the tests from within your Docker image should be implemented. Then you can discuss if you want to implement the automated tests on any other platform.

**Comment by lmoresi** [comment]
Thanks - I had in mind the idea that we would run the tests in a docker environment with all the necessary dependencies which seems to be a documented workflow for travis. This replaces the build-or-die that we currently have with the automated docker build running the tests as the last command in the dockerfile and then flagging a fail-to-build on dockerhub !









Prof Louis Moresi

louis.moresi@unimelb.edu.au<mailto:louis.moresi@unimelb.edu.au>

(w) +61 3 8344 1217

(m) +61 4 0333 1413

(us) +1 505 349 4425

www.moresi.info<http://www.moresi.info/>

www.facebook.com/underworldcode<http://www.facebook.com/underworldcode>

@LouisMoresi<https://twitter.com/LouisMoresi>


On 5 Jun 2019, 10:32 PM +1000, Santiago Soler <notifications@github.com>, wrote:

Configuring Travis to run the tests from within your Docker image would be nice, specially if that's your best recommended way to run Stripy.

My conda proposal was just to run the tests for the non-Docker installation. You could either use the standard Python distribution included on Travis CI VMs, but @brmather<https://github.com/brmather> exposed some issues regarding a few dependencies (gdal, cartopy, lavavu). That's why I offered the conda alternative.

In summary, I agree with @lmoresi<https://github.com/lmoresi>: running the tests from within your Docker image should be implemented. Then you can discuss if you want to implement the automated tests on any other platform.

**Comment by brmather** [comment]
In light of potential issues on other platforms #38 - I think it's we should use Travis CI to test on Linux and OSX. I've created a pull request #40 to update `.travis.yml` configuration file. Amongst other changes, I have removed these optional dependency tests from `test_0_imports.py`:

```python
def test_documentation_dependencies():
    import matplotlib
    import cartopy
    import imageio
    import lavavu
    import pyproj

def test_litho1pt0():
    import litho1pt0
```

A lot of users might not want to install the optional dependencies and just want to install stripy to their system directory, in which case I think we ought to integrate testing on multiple architectures. If they want to run the examples, then they can run the Docker image or Binder. The installation of all the optional dependencies is not a trivial feat!
